src/main/java/org/openjdk/jextract/impl/NameMangler.java

```diff
@@ -136,42 +140,46 @@ public Void visitFunction(Declaration.Function func, Declaration parent) {
         for (Declaration.Variable param : func.parameters()) {
             Type.Function f = Utils.getAsFunctionPointer(param.type());
             if (f != null) {
-                String declFiName = func.name() + "$" + (param.name().isEmpty() ? "x" + i : param.name());
-                JavaFunctionalInterfaceName.with(param, declFiName);
+                String fiName = func.name() + "$" + (param.name().isEmpty() ? "x" + i : param.name());
+                JavaFunctionalInterfaceName.with(param, fiName);
                 i++;
             }
             JavaName.with(param, makeJavaName(param));
+            Utils.forEachNested(param, s -> s.accept(this, func));
         }
+
         Type.Function returnFunc = Utils.getAsFunctionPointer(func.type().returnType());
         if (returnFunc != null) {
             JavaFunctionalInterfaceName.with(func, func.name() + "$return");
         }
+        Utils.forEachNested(func, s -> s.accept(this, func));
 
         return null;
     }
 
     @Override
     public Void visitScoped(Declaration.Scoped scoped, Declaration parent) {
-        String name = scoped.name();
-        if (name.isEmpty() && parent != null) {
-            name = parent.name();
-        }
-        if (JavaName.isPresent(scoped)) {
-            //skip struct that's seen already
-            return null;
-        }
+        if (Utils.isEnum(scoped)) {
+            scoped.members().forEach(fieldTree -> fieldTree.accept(this, null));
```

Comment:
for enums, we don't have to create a new named scope. Also, enums are the only case where jextract leaves an anon enum at the toplevel (where parent == null), so better to skip.